File: backend/app/api/schemas/scan.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ScanSubmit(BaseModel):
    qr_code: str = Field(min_length=1, max_length=500)


class TierProgress(BaseModel):
    quarter: str
    current_count: int
    next_tier: str | None
    scans_required: int | None


class CompEarned(BaseModel):
    amount: float
    type: str
    lifetime_comps: float
    wallet_balance: float


class ScanResponse(BaseModel):
    success: bool
    product_name: str
    usdt_earned: float
    tier_multiplier: float
    tier_progress: TierProgress
    comp_earned: CompEarned | None
    milestone_hit: bool
    wallet_balance: float
    global_scan_count: int


class ScanHistoryItem(BaseModel):
    id: uuid.UUID
    product_name: str | None
    scanned_at: datetime

    model_config = {"from_attributes": True}


class ScanHistoryPage(BaseModel):
    items: list[ScanHistoryItem]
    total: int
    page: int
    per_page: int


class QRCodeResponse(BaseModel):
    id: uuid.UUID
    product_code: str
    unique_id: str
    full_code: str
    is_used: bool
    scanned_by: uuid.UUID | None
    scanned_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class QRCodeGenerateRequest(BaseModel):
    product_id: uuid.UUID
    quantity: int = Field(ge=1, le=10000)


class QRCodeGenerateResponse(BaseModel):
    generated: int
    codes: list[str]


class QRCodeListPage(BaseModel):
    items: list[QRCodeResponse]
    total: int
    page: int
    per_page: int

File: ios/BlakJaks/Features/ScanWallet/Models/Scan.swift
import Foundation

struct Scan: Codable, Identifiable {
    let id: Int
    let qrCode: String
    let productName: String
    let productSku: String
    let usdcEarned: Double
    let tierMultiplier: Double
    let tier: String
    let createdAt: String
}

struct ScanResult: Codable {
    let success: Bool
    let productName: String
    let usdcEarned: Double
    let tierMultiplier: Double
    let tierProgress: TierProgress
    let compEarned: CompEarned?
    let milestoneHit: Bool
    let walletBalance: Double
    let globalScanCount: Int
}

struct TierProgress: Codable {
    let quarter: String
    let currentCount: Int
    let nextTier: String?
    let scansRequired: Int?
}

struct CompEarned: Codable {
    let id: String
    let amount: Double
    let status: String
    let requiresPayoutChoice: Bool
}

struct CompVault: Codable {
    let availableBalance: Double
    let lifetimeComps: Double
    let goldChips: Int
    let milestones: [CompMilestone]
    let guaranteedComps: [GuaranteedComp]
}

struct CompMilestone: Codable, Identifiable {
    let id: String
    let threshold: Double
    let label: String
    let achieved: Bool
    let achievedAt: String?
}

struct GuaranteedComp: Codable, Identifiable {
    let id: Int
    let amount: Double
    let month: String
    let status: String
    let paidAt: String?
}
